pkg/controller/workloadspread/workloadspread_controller.go

```diff
@@ -0,0 +1,713 @@
+		return r.getPodDeployment(targetRef, ws.Namespace)
+	case controllerKindRS.Kind:
+		scaleSelector, err = r.getPodReplicasSet(targetRef, ws.Namespace)
+	case controllerKruiseKindCS.Kind:
+		scaleSelector, err = r.getPodKruiseCloneSet(targetRef, ws.Namespace)
+	case controllerKindJob.Kind:
+		scaleSelector, err = r.getPodJob(targetRef, ws.Namespace)
+	default:
+		r.recorder.Eventf(ws, corev1.EventTypeWarning,
+			"TargetReferenceError", "targetReference is not been recognized")
+		return nil, -1, nil
+	}
+
+	if err != nil {
+		klog.Errorf("WorkloadSpread (%s/%s) handles targetReference failed: %s", ws.Namespace, ws.Name, err.Error())
+		return nil, -1, err
+	}
+	if scaleSelector == nil || scaleSelector.selector == nil || scaleSelector.UID == "" {
+		klog.Errorf("WorkloadSpread (%s/%s) targetReference cannot be found", ws.Namespace, ws.Name)
+		return nil, 0, nil
+	}
+
+	labelSelector, err := metav1.LabelSelectorAsSelector(scaleSelector.selector)
+	if err != nil {
+		klog.Errorf("WorkloadSpread (%s/%s) gets labelSelector failed: %s", ws.Namespace, ws.Name, err.Error())
+		return nil, -1, nil
+	}
+
+	podList := &corev1.PodList{}
+	listOption := &client.ListOptions{
+		Namespace:     ws.Namespace,
+		LabelSelector: labelSelector,
+		FieldSelector: fields.SelectorFromSet(fields.Set{fieldindex.IndexNameForOwnerRefUID: string(scaleSelector.UID)}),
+	}
+	err = r.List(context.TODO(), podList, listOption)
+	if err != nil {
+		return nil, -1, err
+	}
+
+	matchedPods := make([]*corev1.Pod, 0, len(podList.Items))
+	for i := range podList.Items {
+		matchedPods = append(matchedPods, &podList.Items[i])
+	}
+	return matchedPods, scaleSelector.scale, nil
+}
+
+// syncWorkloadSpread is the main logic of the WorkloadSpread controller. Firstly, we get Pods from workload managed by
+// WorkloadSpread and then classify these Pods to each corresponding subset. Secondly, we set Pod deletion-cost annotation
+// value by compare the number of subset's Pods with the subset's maxReplicas, and then we consider rescheduling failed Pods.
+// Lastly, we update the WorkloadSpread's Status and clean up scheduled failed Pods. controller should collaborate with webhook
+// to maintain WorkloadSpread status together. The controller is responsible for calculating the real status, and the webhook
+// mainly counts missingReplicas and records the creation or deletion entry of Pod into map.
+func (r *ReconcileWorkloadSpread) syncWorkloadSpread(ws *appsv1alpha1.WorkloadSpread) error {
+	pods, workloadReplicas, err := r.getPodsForWorkloadSpread(ws)
+	if err != nil || workloadReplicas == -1 {
+		if err != nil {
+			klog.Errorf("WorkloadSpread (%s/%s) gets matched pods failed: %v", ws.Namespace, ws.Name, err)
+		}
+		return err
+	}
+	if len(pods) == 0 {
+		klog.Warningf("WorkloadSpread (%s/%s) has no matched pods, target workload's replicas[%d]", ws.Namespace, ws.Name, workloadReplicas)
+	}
+
+	// group Pods by subset and the extra pods called orphan.
```

Comment:
It is recommended to add some comments and adjust the function name to facilitate a better understanding of the method's role. Especially for podMap, we can explain it briefly and give an example